src/coreclr/src/tools/crossgen2/ILCompiler.Reflection.ReadyToRun/ReadyToRunReader.cs

```diff
@@ -449,12 +449,46 @@ private void EnsureMethods()
                 bool[] isEntryPoint = new bool[nRuntimeFunctions];
 
                 // initialize R2RMethods
-                ParseMethodDefEntrypoints(isEntryPoint);
+                ParseMethodDefEntrypoints((section, reader) => ParseMethodDefEntrypointsSection(section, reader, isEntryPoint));
                 ParseInstanceMethodEntrypoints(isEntryPoint);
                 CountRuntimeFunctions(isEntryPoint);
             }
         }
 
+        private Dictionary<int, ReadyToRunMethod> _runtimeFunctionToMethod = null;
+
+        private void EnsureEntrypointRuntimeFunctionToReadyToRunMethodDict()
+        {
+            EnsureMethods();
+
+            if (_runtimeFunctionToMethod == null)
+            {
+                _runtimeFunctionToMethod = new Dictionary<int, ReadyToRunMethod>();
+                foreach (var section in _methods)
+                {
+                    foreach (var method in section.Value)
+                    {
+                        if (!_runtimeFunctionToMethod.ContainsKey(method.EntryPointRuntimeFunctionId))
+                            _runtimeFunctionToMethod.Add(method.EntryPointRuntimeFunctionId, method);
+                    }
+                }
+            }
+        }
+
+        public Dictionary<TMethod, ReadyToRunMethod> GetCustomMethodToRuntimeFunctionMapping<TType, TMethod, TGenericContext>(IR2RSignatureTypeProvider<TType, TMethod, TGenericContext> provider)
```

Comment:
Can we make this an `IReadOnlyDictionary` to keep this public interface immutable?
We should not allow changing the dictionary anyway.

Reply:
Specifically, this Dictionary isn't cached anywhere, and actually *is* safely mutable, but keeping the public api as providing an immutable surface isn't unreasonable. I'll make that change.